### PR TITLE
Store grid view selection in url params

### PR DESCRIPTION
**Pull request by bbovenzi**
Store the selected dag run and task instance with `dag_run_id` and `task_id` url params for Grid view.

This allowed us to simplify `useSelection` to no longer need useContext. Then I moved some non-components from `/tree` to `/tree/utils`

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by bbovenzi** [comment]
CC: @staufferl16

File: airflow/www/static/js/tree/api/useTasks.js [hunk 1]
```diff
@@ -27,5 +27,8 @@ export default function useTasks(dagId) {
   return useQuery(
     ['tasks', dagId],
     () => axios.get(tasksUrl),
+    {
+      placeholderData: { tasks: [] },
```
Comment (staufferl16) [inline airflow/www/static/js/tree/api/useTasks.js:31]: Is this argument a default return value for the `useQuery` function?

**Review by staufferl16 — APPROVED** [review]
Router/search param code looks great to me!  I've been using MemoryRouter for tests too.

File: airflow/www/static/js/tree/api/useTasks.js (hunk 1 above)
Comment (bbovenzi) [reply airflow/www/static/js/tree/api/useTasks.js:31]: Yes, but it doesn't affect the cache https://react-query.tanstack.com/guides/placeholder-query-data

**Review by pierrejeambrun — APPROVED** [review]
LGTM :+1:

File: airflow/www/static/js/tree/dagRuns/index.test.jsx [hunk 2]
```diff
@@ -42,13 +42,13 @@ const Wrapper = ({ children }) => {
           <ContainerRefProvider value={{}}>
             <TimezoneProvider value={{ timezone: 'UTC' }}>
               <AutoRefreshProvider value={{ isRefreshOn: false, stopRefresh: () => {} }}>
-                <SelectionProvider value={{ onSelect: () => {}, selected: {} }}>
+                <MemoryRouter>
```
Comment (ashb) [inline airflow/www/static/js/tree/dagRuns/index.test.jsx:45]: Why do we use MemoryRouter here but BrowserRouter elsewhere?

Does MemoryRouter mean that components here don't change the URL, right?

Comment (ashb) [reply airflow/www/static/js/tree/dagRuns/index.test.jsx:45]: Oh. This is a `test.jsx` :D

**Review by ashb — APPROVED** [review]

File: airflow/www/static/js/tree/dagRuns/index.test.jsx (hunk 2 above)
Comment (bbovenzi) [reply airflow/www/static/js/tree/dagRuns/index.test.jsx:45]: Yes, tests aren't run on an actual browser so we use a different router.

**Pull request merged**